postgres/src/transaction-manager.ts

```diff
@@ -1,61 +1,65 @@
-    this.#doCommit = (options.prepared ?? true) ? commit : unsafeCommit
+export function getTransactionManager(type = 'default'): TransactionManager {
+  switch (type) {
+    case 'none': return null
+    case 'default': return new DefaultTransactionManager()
+    case 'two-phase': return new TwoPhaseTransactionManager()
+    default: throw TypeError(`Unknown transaction manager type "${type}"`)
   }
+}
 
-  protected async createClient(name: string): Promise<PoolClient> {
-    const client = await super.createClient(name)
+class DefaultTransactionManager implements TransactionManager {
+  async onConnect(client: PoolClient): Promise<void> { // eslint-disable-line class-methods-use-this
     await client.query('BEGIN')
-    return client
   }
 
-  public async transactional<T>(task: () => Promise<T>): Promise<T> {
-    try {
-      const result = await task()
-      await this.commit()
-      return result
-    } catch (err) {
-      await this.rollback()
-      throw err
-    }
+  async onSuccess(clients: PoolClient[]): Promise<void> { // eslint-disable-line class-methods-use-this
+    for (const client of clients) await client.query('COMMIT')
   }
 
-  public async* transactionalYield(yieldValue: any) {
-    try {
-      const result = yield yieldValue
-      await this.commit()
-      return result
-    } catch (err) {
-      await this.rollback()
-      throw err
-    }
+  async onError(clients: PoolClient[]): Promise<void> { // eslint-disable-line class-methods-use-this
+    await Promise.allSettled(clients.map(client => client.query('ROLLBACK')))
   }
+}
 
-  async commit() {
-    try {
-      await this.#doCommit(await this.getClients())
-    } finally {
-      await this.release()
-    }
+class TwoPhaseTransactionManager implements TransactionManager {
+  private transactionsIds = new Map<PoolClient, string>()
+
+  async onConnect(client: PoolClient): Promise<void> { // eslint-disable-line class-methods-use-this
+    await client.query('BEGIN')
   }
 
-  async rollback() {
-    try {
-      await rollback(await this.getClients())
-    } finally {
-      await this.release(true)
+  async onSuccess(clients: PoolClient[]): Promise<void> {
+    // FIXME use Promise.all ?
+    for (const client of clients) {
+      const id = uuid()
+      await client.query(`PREPARE TRANSACTION '${id}'`)
+      this.transactionsIds.set(client, id)
+    }
+
+    // FIXME use Promise.all ?
+    for (const client of clients) {
+      if (!this.transactionsIds.has(client)) throw Error('no transaction id found for client')
+      await client.query(`COMMIT PREPARED '${this.transactionsIds.get(client)}'`)
+      this.transactionsIds.delete(client)
     }
   }
+
+  async onError(clients: PoolClient[]): Promise<void> {
+    await Promise.allSettled(clients.map(async (client) => {
+      if (this.transactionsIds.has(client)) {
+        await client.query(`ROLLBACK PREPARED '${this.transactionsIds.get(client)}'`)
+      } else {
+        await client.query('ROLLBACK')
```

Comment:
There is an edge case here were you have the client have already `COMMIT PREPARED` which can't be rollbacked.

Should we keep track of committed client to avoid trying to rollback them ? Or do we just let it fail, since it's a very rare edge case. If it fails, since we are in a `allSetlled`, it shouldn't have much impact.